inc/az_ulib_ustream_forward_base.h

```diff
@@ -0,0 +1,339 @@
+ * @param[out]      size                The `size_t* const` that points to the place where the
+ *                                      read shall store the number of valid `uint8_t` values
+ *                                      returned in the local buffer.
+ *
+ * @pre     \p ustream_forward          shall not be `NULL`.
+ * @pre     \p ustream_forward          shall be a valid ustream_forward that is the implemented 
+ *                                      ustream_forward type.
+ * @pre     \p buffer                   shall not be `NULL`.
+ * @pre     \p buffer_length            shall not be bigger than 0.
+ * @pre     \p size                     shall not be `NULL`.
+ *
+ * @return The #az_result with the result of the read operation.
+ *      @retval #AZ_OK                        If the ustream_forward copied the content of the 
+ *                                            `Data Source` to the local buffer with success.
+ *      @retval #AZ_ERROR_ULIB_BUSY           If the resource necessary to read the ustream_forward
+ *                                            content is busy.
+ *      @retval #AZ_ERROR_CANCELED            If the read of the content was cancelled.
+ *      @retval #AZ_ULIB_EOF                  If there are no more `uint8_t` values in the `Data
+ *                                            Source` to read.
+ *      @retval #AZ_ERROR_NOT_ENOUGH_SPACE    If there is not enough memory to execute the read.
+ *      @retval #AZ_ERROR_ULIB_SECURITY       If the read was denied for security reasons.
+ *      @retval #AZ_ERROR_ULIB_SYSTEM         If the read operation failed on the system level.
+ */
+AZ_INLINE az_result az_ulib_ustream_forward_read(
+    az_ulib_ustream_forward* ustream_forward,
+    uint8_t* const buffer,
+    size_t buffer_length,
+    size_t* const size)
+{
+  return ustream_forward->_internal.api->read(ustream_forward, buffer, buffer_length, size);
+}
+
+/**
+ * @brief   Returns the size of the ustream_forward.
+ *
+ *  This API returns the number of bytes in the ustream_forward.
+ *
+ *  The `az_ulib_ustream_forward_get_size` API shall follow the following minimum requirements:
+ *      - The API shall return the number of bytes in the ustream_forward.
+ *      - The API shall satisfy all of the precondition requirements laid forth below.
+ *
+ * @param[in]   ustream_forward     The #az_ulib_ustream_forward* with the interface of the 
+ *                                  ustream_forward.
+ * @param[out]  size                The `size_t* const` to return the number of `uint8_t` values.
+ *
+ * @pre     \p ustream_forward      shall not be `NULL`.
+ * @pre     \p ustream_forward      shall be a valid ustream_forward that is the implemented 
+ *                                  ustream_forward type.
+ * @pre     \p size                 shall not be `NULL`.
+ *
+ * @return The #az_result with the result of the `get_size` operation.
```

Comment:
>az_result [](http://example.com/codeflow?start=16&length=9)

I don't think this API can fail. So what if we just return the size instead of az_result?

Reply:
true! I like the idea.

Reply:
so funny enough this actually leads to the most basic implementation of flush having no ability to fail. Let's discuss more tomorrow.